File: cpp/SegmentTree.cpp
constexpr int clog2(int n) { return ((n < 2) ? 1 : 1 + clog2(n / 2)); }

template<typename T>
class SegmentTree
{
public:
    template<typename M>
    SegmentTree(const M& m): merge(m) {}

    void init(vector<T>& raw_)
    {
        raw = raw_;
        n = (int)raw.size();
        int sz = (1 << (clog2(n) + 1));
        data.resize(sz);

        _init(raw, 1, 0, n - 1);
    }

    T modify(int idx, function<T(T)> modifier) { return update(idx, modifier(raw[idx])); }
    T update(int idx, const T& newVal) { raw[idx] = newVal; return _update(1, 0, n - 1, idx, newVal); }
    T query(int left, int right) { return _query(1, 0, n - 1, left, right); }

private:
    vector<T> raw;
    vector<T> data;
    int n;
    using Merge = function<T(const T&, const T&)>;
    Merge merge;

    T _init(vector<T>& raw, int node, int start, int end)
    {
        int mid = (start + end) / 2;
        if (start == end)
            return data[node] = raw[start];
        else
            return data[node] = merge(_init(raw, node * 2, start, mid),
                _init(raw, node * 2 + 1, mid + 1, end));
    }

    T _update(int node, int start, int end, int index, const T& newVal)
    {
        if (index < start || index > end)
            return data[node];

        if (start == end)
            data[node] = newVal;
        else
        {
            int mid = (start + end) / 2;
            data[node] = merge(_update(node * 2, start, mid, index, newVal),
                _update(node * 2 + 1, mid + 1, end, index, newVal));
        }

        return data[node];
    }

    T _query(int node, int start, int end, int left, int right)
    {
        if (left <= start && end <= right)
            return data[node];

        int mid = (start + end) / 2;

        if (mid < left)
            return _query(node * 2 + 1, mid + 1, end, left, right);

        if (mid + 1 > right)
            return _query(node * 2, start, mid, left, right);

        return merge(_query(node * 2, start, mid, left, right),
            _query(node * 2 + 1, mid + 1, end, left, right));
    }
};

File: haskell/SegmentTree.hs
import qualified Data.Sequence as S

data SegTree x = Node { val :: x, segment :: (Int, Int), merge :: (x->x->x), left :: SegTree x, right :: SegTree x } | Empty

isEmpty :: SegTree x -> Bool
isEmpty Empty = True
isEmpty _ = False

query :: SegTree x -> Int -> Int -> x
query (Node v (l,r) merge lc rc) ls rs
    | l == ls && r == rs = v
    | not (isEmpty lc) && ls >= (fst $ segment rc) = query rc ls rs
    | not (isEmpty rc) && rs <= (snd $ segment lc) = query lc ls rs
    | otherwise = let lval = query lc ls (snd $ segment lc)
                      rval = query rc (fst $ segment rc) rs
                  in merge lval rval

update :: SegTree x -> Int -> x -> SegTree x
update (Node v (l,r) merge lc rc) idx nv
    | l == idx && r == idx = Node nv (l,r) merge lc rc 
    | otherwise = (Node (merge (val lc') (val rc')) (l,r) merge lc' rc')
    where rc' = if not (isEmpty lc) && idx >= (fst $ segment rc) then update rc idx nv else rc
          lc' = if not (isEmpty rc) && idx <= (snd $ segment lc) then update lc idx nv else lc

makeSegTree :: S.Seq x -> (x -> x -> x) -> Int -> Int -> SegTree x
makeSegTree xs merge l r
    | S.null xs = Empty
    | S.length xs == 1 = Node (xs `S.index` 0) (l,r) merge Empty Empty
    | otherwise = Node pval (l, r) merge leftSegTree rightSegTree
    where len = (S.length xs + 1) `div` 2
          (xl, xr) = S.splitAt len xs
          leftSegTree = makeSegTree xl merge l (l+len-1)
          rightSegTree = makeSegTree xr merge (l+len) r
          pval = if isEmpty leftSegTree then val rightSegTree 
                 else if isEmpty rightSegTree then val leftSegTree
                 else merge (val leftSegTree) (val rightSegTree)

segFromList :: (x -> x -> x) -> [x] -> SegTree x
segFromList merge xs = makeSegTree (S.fromList xs) merge 0 (length xs - 1)